src/v/datalake/coordinator/state_update.cc

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/coordinator/state_update.h"
+
+#include "datalake/coordinator/state.h"
+#include "datalake/coordinator/translated_offset_range.h"
+
+#include <iterator>
+
+namespace datalake::coordinator {
+
+checked<add_files_update, stm_update_error> add_files_update::build(
+  const topics_state& state,
+  const model::topic_partition& tp,
+  chunked_vector<translated_offset_range> entries) {
+    add_files_update update{
+      .tp = tp,
+      .entries = std::move(entries),
+    };
+    if (!update.can_apply(state)) {
+        return stm_update_error::failed;
+    }
+    return update;
+}
+
+bool add_files_update::can_apply(const topics_state& state) {
+    if (entries.empty()) {
+        return false;
+    }
+    auto prt_state_opt = state.partition_state(tp);
+    if (!prt_state_opt.has_value()) {
+        // No entries at all, this partition hasn't ever added any files.
+        return true;
+    }
+    const auto& prt_state = prt_state_opt.value().get();
+    auto update_range_start_offset = entries.begin()->start_offset;
+    if (
+      prt_state.pending_entries.empty()
+      && !prt_state.last_committed.has_value()) {
+        // No entries at all, this partition hasn't ever added any files.
+        return true;
+    }
+    auto last_offset_state
+      = prt_state.pending_entries.empty()
+          ? prt_state.last_committed.value()
+          : kafka::offset{prt_state.pending_entries.back().last_offset()};
+    if (last_offset_state() + 1 == update_range_start_offset()) {
+        // The last offset for the partition aligns exactly with where we're
+        // adding new data.
+        return true;
+    }
+    // Misalignment, likely because the current `state` doesn't match the state
+    // the update was built with.
+    return false;
```

Comment:
related comment (doesn't have to be addressed in the PR, but still relevant to overall coordinator stm design):

Iceberg could be enabled for sometime, disabled for a couple of days and enabled again.. Currently translator just fetches the "last committed offset" and resumes from that point on but that data may already be gone  in which case a misalignment is possible and in such cases there should be a way to "force" update the state to reset.. or  perhaps the coordinator state machine should listen for "disable" events and reset this state so the translators can automatically begin from the current LSO of the data partition.

Reply:
I was thinking in that case the translator could would just use `min(last_committed_offset, lwm)` as the start offset, but report it to the coordinator as starting as `last_committed_offset` to ensure the contiguous range. Unless we want explicit manual intervention, that seems like the cleanest way forward? And it seems to match Kafka consumer semantics

Reply:
Ya the problem I think is if the `last_committed_offset` (lco) is still around, so we'd want to skip (lco, lso] for translation, so we have to track that bit of information somewhere (either on the data partition or coordinator)?

I was thinking it'd be coordinator because we'd want to GC this in memory state map for (iceberg) disabled topics anyway (we never know if iceberg will ever be enabled again on that topic) ? So in that process just reset this bit? That way this information is centralized in the coordinator topic and the translator is lean and reconcile on demand and coordinator remains the source of truth.

Reply:
Hm good question. In my mind actually we'd just rely on _local_ lwm, to "skip" (lco, lso] roughly.

Reply:
we discussed briefly offline, the idea is to use local start offset of the log to resume translation if there is no check pointed state, this means gaps are possible depending on the sequence of iceberg enable/disable commands and how spaced out they are, so there has to be some gap management built into the coordinator (could be as simple as gap detection and logging to begin with).